### ISSUE TITLE
laya_finetune_typed_decisions_2xT4_kaggle.ipynb fine-tuning task is supervised learning, not an RL problem

### ISSUE THREAD

**Issue by pacman100** [NandhaKishorM/laya#238]
## This fine-tuning task is supervised learning, not an RL problem

I reviewed the [typed-decisions fine-tuning notebook](https://github.com/NandhaKishorM/laya/blob/main/notebooks/laya_finetune_typed_decisions_2xT4_kaggle.ipynb). 

As posed in this notebook, the training problem is fully supervised. Each input comes with a gold probability distribution, and the reward is calculated by comparing the model’s prediction with that known distribution. The model does not act in an environment to discover a reward; there are no state transitions or delayed outcomes to assign credit to. Sampling noisy predictions and applying a policy-gradient estimator changes **how the gradient is calculated**, not what information the training problem provides.

In the [notebook](https://github.com/NandhaKishorM/laya/blob/main/notebooks/laya_finetune_typed_decisions_2xT4_kaggle.ipynb), `gold["probabilities"]` becomes `target`. The `proper_reward` function then scores predicted probabilities against that target using differentiable terms. The training loop detaches the logits, evaluates reward under `no_grad`, and uses normalized advantages in a REINFORCE-style loss alongside soft-target cross entropy. Those gradient barriers are introduced by the implementation; the scoring objective itself does not require them.

For example, the sampled scoring term can be written as a directly differentiable supervised loss:

```python
k = mask.sum(-1, keepdim=True).float()
eps = torch.randn((4,) + logits.shape, device=logits.device) * sigma * mask
eps = (eps - eps.sum(-1, keepdim=True) / k) * mask

q = torch.softmax(
    (logits.unsqueeze(0) + eps).masked_fill(~mask, -1e4),
    dim=-1,
)

loss_score = -proper_reward(
    q, target.unsqueeze(0), qtype, mask,
    w_sph=0.75, w_rps=1.0,
).mean()
loss_ce = -(
    target * torch.log_softmax(
        logits.masked_fill(~mask, -1e4), dim=-1
    )
).sum(-1).mean()

loss = loss_ce + loss_score
```

This retains the noisy predictions, known targets, scoring rule, and cross-entropy term while backpropagating through the reward. It targets the same expected noisy scoring objective as an *unnormalized* Gaussian score-function estimator. Note that it does **not** reproduce the notebook’s exact updates: the four-sample baseline includes each sample’s own reward, and the subsequent batch-wide advantage normalization changes the gradient.

I am not arguing that REINFORCE cannot be applied here. I am questioning whether calling this an **RL problem** accurately describes a task with complete supervised targets and an analytically differentiable reward. A more precise description would be *supervised fine-tuning using a policy-gradient estimator*.

Could you compare cross entropy alone, directly differentiated scoring rules, the reparameterized noisy scoring loss above, and the current Could you compare, under matched training conditions:

Soft-target cross entropy alone;
Cross entropy plus directly differentiated proper scoring;
Cross entropy plus the reparameterized noisy scoring objective above; and
The current policy-gradient method?

Accuracy, Brier score, ECE, training stability, and runtime would make the trade-off clear. If RLCD is intended primarily for future non-differentiable or externally supplied rewards, clarifying that distinction in the notebook and README would also help.

**Comment by NandhaKishorM** [comment]
Thank you, you are right, and this is a clear write-up. Each item comes with a gold distribution, and the reward is a differentiable function of the predicted probabilities, so sampling noise and a policy-gradient estimator only make the gradient noisier. It adds no information, and the extra noise grows with the number of options, which is where laya is weakest. The direct form you wrote is the better loss. A PR that switches the typed-decisions notebook to it, with a before/after on the notebook's held-out split, would be very welcome.

**Comment by liwenjie200543** [comment]
The code-level part of this checks out against the notebook: `gold["probabilities"]` becomes
`target` in the training items, and `proper_reward` scores predictions against it — the training
signal is fully known rather than discovered by acting in an environment.

I can't take on the four-way comparison — CE alone / CE + directly differentiated scoring / CE +
the reparameterized noisy scoring loss / the current policy-gradient method. That's a training
campaign under matched conditions rather than a patch, and I don't have the compute to run it.

What I can do is the documentation half of your last paragraph: make the supervision signal
explicit where the RLCD recipe is named, without renaming the method or touching code. Three places
currently describe RLCD by its gradient mechanism only:

- `README.md`, Fine-Tuning section
- `docs/finetune_browser_agent.md`, step 7 of the recipe
- the §4 intro cell of the typed-decisions notebook

I'm starting with only the `docs/finetune_browser_agent.md` line, because #53 rewrites that README
paragraph and deletes the §4 notebook cell, and I'd rather not open something that conflicts with
it. Happy to extend the wording to those two once #53 lands, or to drop the change entirely if you
prefer RLCD described differently — @NandhaKishorM, your call on the framing as well: if "RL
problem" is the description you want for this objective, say so and I'll leave it alone.

**Comment by liwenjie200543** [comment]
Following up, because your reply landed about 90 seconds before mine and I hadn't seen it when I
posted — so I framed the wrong ask. You narrowed it more usefully than I did: switch the
typed-decisions notebook to the direct form and show a before/after on its held-out split, not the
four-way comparison I said I couldn't run. I'll take that on.

Two practical notes before I start:

- A real before/after means running the notebook twice on Kaggle's 2×T4 (the reference run is 4–5 h
  each, and I don't want to hand you a code-only diff as evidence), so expect a couple of days
  rather than one sitting.
- #53 rewrites this exact notebook — it moves the loop into `laya.finetune` and drops the inline DDP
  script — and its README paragraph names the same reward. Do you want the direct loss in the
  notebook, in `laya.finetune`, or both? If it's cleaner for you, I'll wait for #53 to land and
  rebase onto it instead of racing it.

Separately, the PR I opened (#247) only adds the gold-distribution target to the recipe line in
`docs/finetune_browser_agent.md`; if the notebook change makes that wording obsolete, say so and
I'll drop it.

**Comment by NandhaKishorM** [comment]
Thank you @liwenjie200543. #247, now merged, names the gold-distribution target in the recipe line. The four-way comparison stays open for anyone with the compute, and the notebook's held-out split is a good place to report it.
